From 8e9e46f6ed93ea6d1ebb52b243bf79af19054543 Mon Sep 17 00:00:00 2001
From: Rhed Jao <rhedjao@google.com>
Date: Tue, 3 Jan 2023 10:49:58 +0530
Subject: [PATCH] Fix permanent denial of service via
 setComponentEnabledSetting

Do not update invalid component enabled settings to prevent the
malicious apps from exhausting system server memory.

Bug: 240936919
Test: atest android.security.cts.PackageManagerTest
(cherry picked from commit 5e98f267592775a2b886ccaa752377d6967f9741)
Merged-In: I08165337895e89f13a2b9fcce1201cba9ad13d7d
---
 .../core/java/com/android/server/pm/PackageManagerService.java | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/services/core/java/com/android/server/pm/PackageManagerService.java b/services/core/java/com/android/server/pm/PackageManagerService.java
index cde249fe2a72..2922c56fb47c 100644
--- a/services/core/java/com/android/server/pm/PackageManagerService.java
+++ b/services/core/java/com/android/server/pm/PackageManagerService.java
@@ -21137,6 +21137,9 @@ public class PackageManagerService extends IPackageManager.Stub
                     } else {
                         Slog.w(TAG, "Failed setComponentEnabledSetting: component class "
                                 + className + " does not exist in " + packageName);
+			// Safetynet logging for b/240936919
+			EventLog.writeEvent(0x534e4554, "240936919", callingUid);
+			return;
                     }
                 }
                 switch (newState) {
-- 
2.17.1

